🪪 feat: Microsoft Graph Access Token Placeholder for MCP Servers

maxesse opened the pull request:
## Summary

Adds support for a new `{{LIBRECHAT_GRAPH_ACCESS_TOKEN}}` placeholder in MCP server configurations.

The existing `{{LIBRECHAT_OPENID_ACCESS_TOKEN}}` provides the user's raw OIDC token, which is audience-scoped to LibreChat and cannot be used to call Microsoft Graph API directly. This new placeholder performs an OAuth On-Behalf-Of (OBO) exchange to obtain a token that Microsoft Graph API will accept, enabling MCP servers to make Graph API calls on behalf of the authenticated user.

The main use case is for M365-related MCP servers such as the M365 MCP by Softeria, which support bypassing the full OAuth flow and just perform requests with an externally managed graph token. This improves the user experience by bypassing an additional OAuth flow when the same token is already present (when LibreChat is configured for OIDC token reuse and OBO flows).

**Changes:**
- New `graph.ts` utility with core logic for detecting and resolving Graph token placeholders in MCP options (headers, env vars, URLs)
- New OIDC constants: `GRAPH_TOKEN_PLACEHOLDER` and `DEFAULT_GRAPH_SCOPES`
- MCPManager integration to pre-process Graph token placeholders asynchronously before the synchronous `processMCPEnv` pipeline
- MCP service wiring to inject `graphTokenResolver` from `GraphTokenService` into tool instances

**Usage:**
```json
{
  "headers": {
    "Authorization": "Bearer {{LIBRECHAT_GRAPH_ACCESS_TOKEN}}"
  }
}
```

Scopes can be configured via the GRAPH_API_SCOPES environment variable (defaults to https://graph.microsoft.com/.default).

## Change Type

Please delete any irrelevant options.

- [x] New feature (non-breaking change which adds functionality)

## Testing

1. Configure an MCP server with {{LIBRECHAT_GRAPH_ACCESS_TOKEN}} in headers
2. Authenticate via OpenID Connect with a Microsoft Entra ID provider
3. Invoke a tool from the configured MCP server
4. Verify the Graph token is resolved and the MCP server receives a valid token

### **Test Configuration**:

- OpenID Connect configured with Microsoft Entra ID
- GRAPH_API_SCOPES environment variable (optional, defaults to https://graph.microsoft.com/.default)

## Checklist

Please delete any irrelevant options.

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] My changes do not introduce new warnings
- [x] Local unit tests pass with my changes

maxesse committed:
feat: MCP Graph Token env var

maxesse commented:
That's right @danny-avila, it's basically the same but we do the OBO flow on the Librechat side, instead of leaving that task to the mcp at the other end.

maxesse commented:
@danny-avila just pushed the docs change PR, covers both this one and #9931 🎉

maxesse commented on packages/api/src/utils/graph.ts:
```diff
@@ -0,0 +1,206 @@
+    return options;
+  }
+
+  const result = { ...options };
```
The shallow copy is intentionally sufficient here. The implementation replaces the entire headers and env properties with new objects returned from resolveGraphTokensInRecord(), rather than mutating nested properties. This means the original options.headers and options.env remain untouched. If nested mutation becomes necessary in the future, we can revisit this approach.

maxesse committed:
Addressing copilot remarks

maxesse committed:
Addressed Copilot review remarks

maxesse commented:
All done :) I applied all of copilot's changes (except one - see comment), the merge conflic is solved.

maxesse commented:
Yeah I'll take a look this evening when I get home - I looked at it quickly the other day, it was mostly related to mcp.js and the axios mock, not sure it was related to this change but I need to read through it with more calm, I'm sure I can fix it.

maxesse committed:
Fixed graphtokenservice mock in MCP test suite

maxesse commented:
@danny-avila test passing now 🤌🏻

maxesse committed:
fix: remove unnecessary type check and cast in resolveGraphTokensInRecord

danny-avila commented:
Changed base to `dev-staging` branch since this will allow changes to the codebase before the next releast, but intended for the next release `v0.8.3-rc1`, there are some merge conflicts now given necessary changes, will review this week

danny-avila committed:
ci: add Graph Token integration tests in MCPManager

danny-avila committed:
refactor: update user type definitions to use Partial<IUser> in multiple functions

danny-avila committed:
test: enhance MCP tests for graph token processing and user placeholder resolution

- Added comprehensive tests to validate the interaction between preProcessGraphTokens and processMCPEnv.
- Ensured correct resolution of graph tokens and user placeholders in various configurations.
- Mocked OIDC utilities to facilitate testing of token extraction and validation.
- Verified that original options remain unchanged after processing.

danny-avila committed:
chore: import order

danny-avila committed:
chore: imports
